src/portfft/dispatcher/workitem_dispatcher.hpp

```diff
@@ -141,24 +145,67 @@ PORTFFT_INLINE void workitem_impl(const T* input, T* output, const T* input_imag
   auto loc_load_modifier_view = detail::padded_view(loc_load_modifier, BankLinesPerPad);
   auto loc_store_modifier_view = detail::padded_view(loc_store_modifier, BankLinesPerPad);
 
-  for (IdxGlobal i = global_id; i < round_up_to_multiple(n_transforms, static_cast<IdxGlobal>(SubgroupSize));
-       i += global_size) {
-    bool working = i < n_transforms;
-    Idx n_working = sycl::min(SubgroupSize, static_cast<Idx>(n_transforms - i) + subgroup_local_id);
-
-    IdxGlobal global_offset = static_cast<IdxGlobal>(n_io_reals) * (i - static_cast<IdxGlobal>(subgroup_local_id));
-    if (LayoutIn == detail::layout::PACKED) {
-      if (storage == complex_storage::INTERLEAVED_COMPLEX) {
-        global_data.log_message_global(__func__, "loading non-transposed data from global to local memory");
-        global2local<level::SUBGROUP, SubgroupSize>(global_data, input, loc_view, n_reals * n_working, global_offset,
-                                                    local_offset);
+  const IdxGlobal transform_idx_begin = static_cast<IdxGlobal>(global_data.it.get_global_id(0));
+  const IdxGlobal transform_idx_step = static_cast<IdxGlobal>(global_data.it.get_global_range(0));
+  const IdxGlobal transform_idx_end = round_up_to_multiple(n_transforms, static_cast<IdxGlobal>(SubgroupSize));
+  for (IdxGlobal i = transform_idx_begin; i < transform_idx_end; i += transform_idx_step) {
+    const bool working = i < n_transforms;
+    IdxGlobal leader_i = i - static_cast<IdxGlobal>(subgroup_local_id);
+
+    Idx n_working = sycl::min(SubgroupSize, static_cast<Idx>(n_transforms - leader_i));
+    IdxGlobal global_offset = static_cast<IdxGlobal>(n_io_reals) * leader_i;
+    IdxGlobal global_input_offset = static_cast<IdxGlobal>(input_distance_in_reals) * leader_i;
+    IdxGlobal global_output_offset = static_cast<IdxGlobal>(output_distance_in_reals) * leader_i;
+
+    // This is checking for LayoutIn is PACKED or UNPACKED but we don't actually ever launch kernels with LayoutIn
+    // as UNPACKED.
+    if (LayoutIn == detail::layout::PACKED || LayoutIn == detail::layout::UNPACKED) {
+      // copy into local memory cooperatively as a subgroup, allowing coalesced memory access for when elements of a
+      // single FFT are sequential. BATCH_INTERLEAVED skips this step and loads straight from global to registers since
+      // the sequential work-items already access sequential elements.
+      if (input_stride == 1 && input_distance == fft_size) {
```

Comment:
I still don't like that we nest different `if` conditions that test the layout. The first `if (LayoutIn == detail::layout::PACKED || LayoutIn == detail::layout::UNPACKED)` would make sense if the 2 layouts had some common logic but there is almost nothing currently. It would make more sense to reduce the nested-ness like:
```
if (layout_in == PACKED) {
...
} else if (layout_in == BATCH_INTERLEAVED) {
...
} else {
...
}
```
That would only work if you already make the change to define `layout_in` inside the kernel.

Reply:
I was just not keen because it would take more time, but if we have time then I'll happily do it.

Reply:
Sure, addressed in this commit https://github.com/codeplaysoftware/portFFT/pull/136/commits/f8a8661520105d38262d149a3686e87054807ba7, but its probably easier to look at the diff from main